### ISSUE TITLE
How does implicit deal with test data?

### ISSUE THREAD

**Issue by fabianschan** [benfred/implicit#210]
If I split a set and train with the training set. How do you can make recommendations with new users? How can we transform the data of a user (e. g. test_user  = [0,0,0,1,1,0,1]) into user vectors without recomputing the whole model again?

The recommenderbase can handle users of a testset, but how?

**Comment by benfred** [comment]
For cross validation - we don't totally exclude the user from the training set. Instead we split on user-item interactions, so the user will have rated items in both the test and train set. We are then measuring how well the model will rank the items in the test set for a useer, given different items in the training set.

For new users in production - there is a 'recalculate_user' parameter on the recommend call for some models, which can be used for generating recommendations without recalculating the whole model.

**Comment by fabianschan** [comment]
`    def _user_factor(self, userid, user_items, recalculate_user=False):
        if recalculate_user:
            return self.recalculate_user(userid, user_items)
        else:
            return self.user_factors[userid]

    def recalculate_user(self, userid, user_items):
        raise NotImplementedError("recalculate_user is not supported with this model")`

Sorry to open this again, but It seems like we cannot recalculate a new user with ALS

**Comment by benfred** [comment]
Did you try this? It should work for ALS (but not BPR currently).

the recalculate_user method is overridden here:

https://github.com/benfred/implicit/blob/6b16c50d1d514a814f2e5b8cf2a829ff23dbba63/implicit/als.py#L216-L219

which calculates factors for the user here:

https://github.com/benfred/implicit/blob/6b16c50d1d514a814f2e5b8cf2a829ff23dbba63/implicit/als.py#L347-L350

This isn't the fastest code, but should work.

**Comment by fabianschan** [comment]
Wow. Thank you very much. That helped a lot.
May I ask you two additional things?

1.) In the function `user_linear_equation(Y, YtY, Cui, u, regularization, n_factors)` you write `# accumulate YtCuPu in b` but I dont see `Pu` coming up in the for loop:

```
 for i, confidence in nonzeros(Cui, u):
        factor = Y[i]

        if confidence > 0:
            b += confidence * factor
        else:
            confidence *= -1

        A += (confidence - 1) * np.outer(factor, factor)
```
could you explain the role of `Pu` in the equations?

2.) Is there a reason doing that in a foor loop and not with matrix operations?

Thank you!

Regards
Fabian
